### PR TITLE
Feature/newautotests

### LINKED ISSUE #1: Add more tests
Duplicated from https://github.com/DoubleDeez/MDFramework/issues/35

We should add more tests (demos) to test everything added in https://github.com/DoubleDeez/MDFramework/issues/17 and all features that existed before https://github.com/DoubleDeez/MDFramework/issues/17.

- Test all sort of property configuration (unreliable, reliable, interval, onchange, join, one shot)
- Add test for buffered values by adding stuff to the buffer before starting the game (Text box that adds any message to the text)

### LINKED ISSUE #8: More advanced automated tests
The automated tests should run all tests multiple times with different configurations. This should be done by having the clients disconnect and restarting the server before they reconnect.

In addtion we should also have some static nodes already existing in the scene that also perform some tests to test the current bug where these nodes are wiped from the replicator when the session ends.

### PR DESCRIPTION
**Pull request by Beider**
New automated tests framework using attributes and with multiple configurations

### CODE REVIEW

File: AutomaticTests/Managers/TestConfigManager.cs [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        return $"[{Type.Name}]";
+    }
+
+}
+
+public enum TestConfigCommands
+{
+    INIT_CONFIG,    // Sent when we start a new config run, causes before disconnect to be run on current config.
+                    // Then disconnect happens, run after disconnect, before connect, then wait and reconnect. Then after connect is run
+    ACK_INITI_CONFIG, // Sent by clients right before they disconnect
```
Comment (DoubleDeez) [inline AutomaticTests/Managers/TestConfigManager.cs]: Typo? `INITI` -> `INIT`

File: AutomaticTests/Managers/TestConfigManager.cs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+    private int SignalsRecieved = 0;
+
+    public TestConfigManager(MDTestManager TestManager)
+    {
+        this.TestManager = TestManager;
+
+        // Load our config files
+        List<Type> ConfigList = MDStatics.FindAllScriptsImplementingAttribute<MDTestConfig>();
+        if (ConfigList.Count == 0)
+		{
```
Comment (DoubleDeez) [inline AutomaticTests/Managers/TestConfigManager.cs]: Indenting is weird here

File: AutomaticTests/Managers/TestConfigManager.cs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        }
+    }
+
+    /// <summary>
+    /// Called on the server, STEP 1 of the reconnect cycle
+    /// </summary>
+    /// <param name="Config">The next config to load</param>
+    private void RunConfig(TestConfig Config)
+    {
+       InvokeBeforeDisconnect();
```
Comment (DoubleDeez) [inline AutomaticTests/Managers/TestConfigManager.cs]: missing a space for indent

File: AutomaticTests/Managers/TestManager.cs [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+    public TestManager(MDTestManager MDTestManager)
+    {
+        this.MDTestMan = MDTestManager;
+        Instance = this;
+
+        // Load our config files
+        List<Type> FoundTests = MDStatics.FindAllScriptsImplementingAttribute<MDTestClass>();
+        if (FoundTests.Count == 0)
+		{
```
Comment (DoubleDeez) [inline AutomaticTests/Managers/TestManager.cs]: weird indents

**Review by DoubleDeez — DISMISSED** [review]

**Comment by Beider** [comment]
I think all of these things were caused by somehow tabs being part of the indents. I fixed it now

**Review by DoubleDeez — APPROVED** [review]

**Pull request merged**